WalletWasabi/Bases/PeriodicRunner.cs

```diff
@@ -1,132 +1,105 @@
 using Microsoft.Extensions.Hosting;
 using System;
-using System.Collections.Generic;
-using System.Linq;
-using System.Text;
 using System.Threading;
 using System.Threading.Tasks;
 using WalletWasabi.Logging;
 
 namespace WalletWasabi.Bases
 {
+	/// <summary>
+	/// <see cref="PeriodicRunner"/> is an extension of <see cref="BackgroundService"/> that is useful for tasks
+	/// that are supposed to repeat regularly.
+	/// </summary>
 	public abstract class PeriodicRunner : BackgroundService
 	{
+		private volatile TaskCompletionSource<bool>? _tcs;
+
 		protected PeriodicRunner(TimeSpan period)
 		{
-			TriggeringCts = new CancellationTokenSource();
-			TriggerLock = new object();
 			Period = period;
-			ResetLastException();
+			ExceptionTracker = new LastExceptionTracker();
 		}
 
-		private CancellationTokenSource TriggeringCts { get; set; }
-		private object TriggerLock { get; }
 		public TimeSpan Period { get; }
-		public Exception LastException { get; set; }
-		public long LastExceptionCount { get; set; }
-		public DateTimeOffset LastExceptionFirstAppeared { get; set; }
 
-		private void ResetLastException()
-		{
-			LastException = null;
-			LastExceptionCount = 0;
-			LastExceptionFirstAppeared = DateTimeOffset.MinValue;
-		}
+		private LastExceptionTracker ExceptionTracker { get; }
 
+		/// <summary>
+		/// Normally, <see cref="ActionAsync(CancellationToken)"/> user-action is called every time that <see cref="Period"/> elapses.
+		/// This method allows to expedite the process by interrupting the waiting process.
+		/// </summary>
+		/// <remarks>
+		/// If <see cref="ExecuteAsync(CancellationToken)"/> is not actually in waiting phase, this method call makes
+		/// sure that next waiting process will be omitted altogether.
+		/// </remarks>
 		public void TriggerRound()
 		{
-			lock (TriggerLock)
-			{
-				TriggeringCts?.Cancel();
-			}
+			// Note: All members of TaskCompletionSource<TResult> are thread-safe and may be used from multiple threads concurrently.
+			_tcs?.TrySetResult(true);
 		}
 
+		/// <summary>
+		/// Abstract method that is called every <see cref="Period"/> or sooner when <see cref="TriggerRound"/> is called.
+		/// </summary>
 		protected abstract Task ActionAsync(CancellationToken cancel);
 
 		/// <inheritdoc />
 		protected override async Task ExecuteAsync(CancellationToken stoppingToken)
 		{
 			while (!stoppingToken.IsCancellationRequested)
 			{
+				_tcs = new TaskCompletionSource<bool>();
+
 				try
 				{
+					// Do user action.
 					await ActionAsync(stoppingToken).ConfigureAwait(false);
-					LogAndResetLastExceptionIfNotNull();
+
+					// Log previous exception if any.
+					LogLastException(ExceptionTracker.LastException);
```

Comment:
Sadly Github doesn't allow me to comment in lines that were not modified and I need to put this comment here.

In the `catch` below we can see `ex is OperationCanceledException || ex is TaskCanceledException`, I think thats redundant because a TaskCanceledException is an OperationCanceledException.

Reply:
Yes.